src/main/java/com/redhat/devtools/lsp4ij/operations/signatureHelp/LSPParameterInfoHandler.java

```diff
@@ -0,0 +1,285 @@
+package com.redhat.devtools.lsp4ij.operations.signatureHelp;
+
+import com.intellij.codeInsight.CodeInsightBundle;
+import com.intellij.codeInsight.CodeInsightSettings;
+import com.intellij.lang.parameterInfo.*;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.util.Pair;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.psi.PsiElement;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.lsp4ij.LanguageServiceAccessor;
+import com.redhat.devtools.lsp4ij.internal.CancellationSupport;
+import org.eclipse.lsp4j.*;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.concurrent.*;
+
+/**
+ * LSP implementation of {@link ParameterInfoHandler} to support
+ * <a href="https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_signatureHelp>LSP 'textDocument/signatureHelp'</a>.
+ */
+public class LSPParameterInfoHandler implements ParameterInfoHandler<LSPSignatureHelperPsiElement, SignatureInformation> {
+
+    // Methods called when the parameter hint popup appears.
+
+    @Override
+    public @Nullable LSPSignatureHelperPsiElement findElementForParameterInfo(@NotNull CreateParameterInfoContext context) {
+        return toLSPSignatureHelperPsiElement(context);
+    }
+
+    @Override
+    public void showParameterInfo(@NotNull LSPSignatureHelperPsiElement psiElement, @NotNull CreateParameterInfoContext context) {
+        CancellationSupport cancellationSupport = new CancellationSupport();
+        try {
+            ProgressManager.checkCanceled();
+
+            // 1. Collect signature help asynchronously
+            // Here the popup hint is not shown, the signature help is invoked
+            SignatureHelpParams params = toSignatureHelpParams(context, SignatureHelpTriggerKind.Invoked);
+            BlockingDeque<Pair<SignatureHelp, LanguageServer>> signatureHelps = new LinkedBlockingDeque<>();
+            CompletableFuture<Void> future = collectSignatureHelp(context, params, signatureHelps, cancellationSupport);
+            // 2. Show the popup hint when signature help is ready.
+            future.thenAccept(a -> {
+                if (!signatureHelps.isEmpty()) {
+                    // Get the first signature help
+                    SignatureHelp signatureHelp = signatureHelps.getFirst().getFirst();
+                    if (signatureHelp != null) {
+                        // Store the current signature help in the Psi element
+                        psiElement.setActiveSignatureHelp(signatureHelp);
+                        // There is a signature help, display it
+                        List<SignatureInformation> signatures = signatureHelp.getSignatures();

```

Comment:
can signatures be null here? or empty?

Reply:
I take care of that. The main idea is "don't trust the server", so we need to take care of this usecase even if spec says that signatures cannot be null